File: sorting/quick_sort.py
#Quick Sort

def quick_sort(lst):

    """
    >>> quick_sort([1, 6, 0, 3, 5, 4, -20, 7])
    [-20, 0, 1, 3, 4, 5, 6, 7]
    """
    quick_sort_helper(lst, 0, len(lst)-1)

def quick_sort_helper(lst, start, end):
    if start < end:
        #base case where if the list provided <= len 1, it is already sorted
        split_point = partition(lst, start, end)

        quick_sort_helper(lst, start, split_point-1)  #left side of split_point
        quick_sort_helper(lst, split_point+1, end)  #right side of split_point


def partition(lst, start, end):

    pivot = lst[start]
    left = start + 1
    right = end

    done = False
    while not done:
        while left <= right and lst[left] <= pivot:
            left = left + 1

        while lst[right] >= pivot and right >= left:
            right = right - 1

        if right < left:
            done = True
        else:
            temp = lst[left]
            lst[left] = lst[right]
            lst[right] = temp

    temp = lst[start]
    lst[start] = lst[right]
    lst[right] = temp

    return right

File: sorting/quickSort.js
// Quick Sort
// Worstcase Runtime: O(n^2) when every pivot is either the smallest or largest integer
// Usual Runtime: O(n log(n)) divde and conquer
// Additional Spacetime Complexity: None; done in place

function quickSort(arr, start, end) {
  if (start >= end) {
    return
  }
  var pIndex = partition(arr, start, end)
  quickSort(arr, start, pIndex-1)
  quickSort(arr, pIndex+1, end)
  return arr
}

function partition(arr, start, end) {
  var pivotVal = arr[end]
  var pIndex = start

  for (var i = start; i < end; ++i) {
    if (arr[i] < pivotVal) {
      swap(arr, i, pIndex )
      pIndex = pIndex + 1
    }
  }
  swap(arr, end, pIndex)
  return pIndex
}

function swap(arr, i, j) {
  var temp = arr[i]
  arr[i] = arr[j]
  arr[j] = temp
}

console.log(quickSort([11,8,14,3,6,2,7],0,6));
//[2, 3, 6, 7, 8, 11, 14]
console.log(quickSort([11,8,14,3,6,2,1, 7],0,7));
//[1, 2, 3, 6, 7, 8, 11, 14]
console.log(quickSort([16,11,9,7,6,5,3, 2],0,7));
